Reservoir coupling: Implement time stepping

hakonhagland opened the pull request:
Builds on #5521 which should be merged first.

Also refactors the original adaptive time stepping method.

bska commented:
> Builds on #5521 which should be merged first.

Okay, in that case I'm going to mark this as "work in progress/draft" to avoid inadvertent merging.

hakonhagland commented:
> I'm going to mark this as "work in progress/draft" to avoid inadvertent merging.

@bska Thanks. I was actually planning to get this merged, since it is becoming difficult to handle all the dependent PRs and keep them up to date. Is this a good idea, or should I wait until the complete reservoir coupling feature is ready?

atgeirr commented:
> Is this a good idea, or should I wait until the complete reservoir coupling feature is ready?

It is a very good idea. Do not wait for the complete feature, unless you expect significant changes to the parts you already did. This patch has grown to become quite large, so finishing and polishing the first PRs in the series so they can be merged should be a high priority.

hakonhagland commented:
> It is a very good idea.

@atgeirr Ok I agree. I have removed the draft state on https://github.com/OPM/opm-simulators/pull/5436 as it is the first in the dependency chain and we should try to merge that first.

blattms commented:
jenkins build this serial please

blattms commented on opm/simulators/flow/ReservoirCouplingMaster.hpp:
```diff
@@ -53,10 +56,12 @@ class ReservoirCouplingMaster {
 
     const Parallel::Communication &comm_;
     const Schedule& schedule_;
+    std::time_t start_date_;  // Master process' simulation start date
```
Is there a specific reason for the type or should we maybe use the same type for start date everywhere? Seems like it is  `boost::gregorian::date` is SimulationTimer.

blattms reviewed (commented):
Review of commit 3564e1d7098d1, files .../flow/ReservoirCoupling*

Do we really need the TimePoint class?

hakonhagland commented:
> Unfortunately, we require compiling without MPI to work.

@blattms How do build locally without MPI? I tried cmake option `-DUSE_MPI=NO` but opm-grid fails to build with error `/usr/bin/ld: cannot find -lMPI::MPI_C: No such file or directory`

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
jenkins build this please

blattms commented:
> > Unfortunately, we require compiling without MPI to work.
> 
> @blattms How do I build locally without MPI? I tried cmake option `-DUSE_MPI=NO` but opm-grid fails to build with error `/usr/bin/ld: cannot find -lMPI::MPI_C: No such file or directory`

Use `-DCMAKE_DISABLE_FIND_PACKAGE_MPI=ON`, but you need to use the same options for all modules.

hakonhagland commented:
jenkins build this please

hakonhagland commented on opm/simulators/flow/ReservoirCouplingMaster.cpp:
```diff
@@ -44,41 +44,133 @@ ReservoirCouplingMaster::ReservoirCouplingMaster(
+            // The slave process will start during the timestep, but not at the beginning
+            slave_elapsed_time = slave_start_date - step_start_date;
+        }
+        suggested_timestep = slave_elapsed_time;
+        step_end_date = step_start_date + suggested_timestep;
```
We update `step_end_date` for each iteration of the loop if the current value is less than the previous value (since if it is not less than that value, the execution will not reach this point). This is to compute the minimum `step_end_date` and the corresponding `suggested_time_step` over all the slaves. So it is not necessarily "the last one that wins", but rather the one that implies the smallest `step_end_date`. I have added a comment in the code to explain this. See the last commit.

hakonhagland commented:
jenkins build this please

blattms commented on opm/simulators/flow/Main.cpp:
```diff
@@ -35,11 +35,26 @@
 #include <opm/simulators/utils/DamarisOutputModule.hpp>
 #endif
 
+#include <iostream>
+#include <fcntl.h>  // for open()
+#include <unistd.h> // for dup2(), close()
+
+#include <iostream>
+#include <fcntl.h>  // for open()
+#include <unistd.h> // for dup2(), close()
+
+#include <iostream>
+#include <fcntl.h>  // for open()
+#include <unistd.h> // for dup2(), close()
```
These are a lot of c-headers. Please use C++ wherever possible.

Maybe we can use the logging system to do this somehow or postpone this to later?

blattms commented on opm/simulators/flow/Main.cpp:
```diff
@@ -115,6 +130,53 @@ Main::~Main()
+        if (dup2(fd, fileno(stdout)) == -1 || dup2(fileno(stdout), fileno(stderr)) == -1) {
+            std::string error_msg = "Slave: Failed to redirect stdout+stderr to " + filename;
+            perror(error_msg.c_str());
+            close(fd);
```
I think we should skip this. Open failed, hence closing will, too.

hakonhagland replied:
> Open failed

Not sure what you mean? If open() failed program would already be terminated at line 160.

blattms commented on opm/simulators/utils/UnsupportedFlowKeywords.cpp:
```diff
@@ -242,9 +242,7 @@ const KeywordValidation::UnsupportedKeywords& unsupportedKeywords()
         {"GRAVDRB", {true, std::nullopt}},
         {"GRAVDRM", {true, std::nullopt}},
         {"GRDREACH", {true, std::nullopt}},
-        {"GRUPMAST", {true, std::nullopt}},
```
These should stay marked as unsupported until we really support this.

blattms commented on opm/simulators/flow/FlowMain.hpp:
```diff
@@ -32,6 +32,9 @@
 #include <opm/simulators/flow/FlowUtils.hpp>
 #include <opm/simulators/flow/SimulatorFullyImplicitBlackoil.hpp>
 
+#if HAVE_MPI
+#define RESERVOIR_COUPLING_ENABLED
+#endif
```
Was this needed to make serial compilation work? Could we hide this somewhere e.g. by providing methods doing nothing in this case are throwing? Might make the code clearer and cleaner.

hakonhagland replied:
Yes this was done to make serial compilation work. I think it is clearer than using `#if HAVE_MPI`. Instead we use `#ifdef RESERVOIR_COUPLING_ENABLED` which make it clearer. I am not sure I understood your comment about hiding it? Did you mean moving it to a designated header file?

blattms reviewed (commented):
I am not finished with my review yet. I will need to think  a bit more about this (especially time stepping and output redirection) and probably discuss my thoughts with you. Hence this is a bit premature, still. 

For now I am just posting the comments that are there for now.

One more MPI related comment:
I saw that you used MPI_AINT and claimed that this should work on most systems for communicating std::size_t. Luckily there is a way to get corresponding MPI types automatically:
```
#include <dune/common/parallel/mpitraits.hh>
....
auto myMPIType = Dune::MPITraits<std::size_t>::getType();
```
This will do the correct thing on all architectures.

hakonhagland commented:
jenkins build this please

hakonhagland commented on opm/simulators/flow/ReservoirCoupling.cpp:
```diff
@@ -0,0 +1,108 @@
+    else {
+        MPI_Comm_create_errhandler(custom_error_handler_slave_, &errhandler);
+    }
+    MPI_Comm_set_errhandler(comm, errhandler);
```
>One would this directly in main or at least way up in the call stack.

Note that these error handlers are set only for the slave-master communicators. So the default error handler still applies to the MPI_COMM_WORLD communicator set up in main. Currently, the slave-master communicators are first defined in `SimulatorFullyImplicitBlackoil.hpp` in init() or run(). And therefore it is currently not convenient to set error handlers for these in `Main.hpp`. 

Can you elaborate on why it is better to use the default error handler for these communicators? I see two advantages with using custom error handlers: Debugging (e.g. setting breakpoints in the handler and getting backtrace), and opportunity to provide more context in the error message.

hakonhagland replied:
> I doubt that this works as errhandler is a local variable.

As I understand, `errhandler` is just an MPI handle (an integer) so MPI will create its internal errhandler structure corrsponding to that handle when we call `MPI_Comm_create_errhandler()` + `MPI_Comm_set_errhandler(comm, errhandler)`. After that we can simply call `MPI_Errhandler_free()` on that handle. The handle itself (the local `errhandler` variable) is just a copy of an integer and can safely be thrown away as the function `setErrhandler()` returns. I tried to clarify this in the latest commit.

blattms replied:
To me these functions are for memory management that needs to be done in your code. The handle is basically a pointer to an internally allocated object.

You do have point, though. There is no lifetime issue as this is just handle and the err handler is allocated internally and lives until `MPI_Errhandler_free` is called on it. As this is never called we just have a memory leak. Not nice, but also not fatal.

But if you call `MPI_Errhandler_free`  in this function (as you proposed) then you will have created a memory problem inside of MPI.

The issue that I see with this is code is that I cannot simply overwrite the error handler anymore in Main.cpp or somewhere else up the stack. You unconditionally reset it here. Hence I would to change it here, too.

hakonhagland replied:
> But if you call MPI_Errhandler_free in this function (as you proposed) then you will have created a memory problem inside of MPI.

After checking the source code of open MPI, I would claim that this is not memory problem inside MPI. Refer to 
[comm_set_errhandler.c](https://github.com/open-mpi/ompi/blob/main/ompi/mpi/c/comm_set_errhandler.c) and [errhandler_free.c](https://github.com/open-mpi/ompi/blob/main/ompi/mpi/c/errhandler_free.c). It shows that the communicator also claims a reference to the errhandler. So it seems correct to call `MPI_Errhandler_free()` immediately. This will release the memory of the errhandler when the communicator is deallocated.

hakonhagland commented:
jenkins build this please

hakonhagland commented:
I was not able to rebase this on master (ref. private discussion with @blattms). (After resolving conflicts the resulting files were messed up and couldn't even be compiled). As a workaround I am merging master into this branch instead. The latest merge commit incorporates the following pull requests that have been merged into master since last time master was merged and which also changed the file `AdaptiveTimeStepping.hpp`: 

- #2970: remove 16 days restriction for timestep in predicition mode
- #5803: remove unused fipnum argument
- #5792: added: support WCYCLE

atgeirr commented:
> As a workaround I am merging master into this branch instead

In my experience that creates an almost impossible mess to extract oneself from. While I do not know what was the reason for the rebasing failures, it can sometimes help to squash the commits on your branch and then rebase, at least then each conflict will only arise once whereas otherwise you may have a conflict that appears over and over again every time your branch commits touch the same code parts. Maybe you could try again after first squishing?

Advice from git experts welcome!

akva2 commented:
This cannot be rebased on master since it's based on the spawn_slaves branch. Rebasing spawn_slaves on master, then this PR on spawn slaves, all seems fine to me.

hakonhagland commented:
> since it's based on the spawn_slaves branch 

@akva2 How did you check that? I tried:

```
$ git branch --show-current
timestepping2
$ git merge-base --is-ancestor spawn_slaves HEAD || echo "Not based on spwan_slaves"
Not based on spwan_slaves
```

so it seems it is not based on that branch?

akva2 commented:
First commit in this branch modifies [ReservoirCouplingMaster.cpp](https://github.com/OPM/opm-simulators/pull/5620/commits/b40f5ec9a2eedbe96876097fd10e36681c455e28) and that file looked to me as is added in https://github.com/hakonhagland/opm-simulators-1/commit/13c2b43ce4ffa1e89ace3e83cb965c578e708b20

akva2 commented:
Yes. You are going about this the wrong way around. 

I told you what the *PROBLEM* is. The problem is that commit *IS NOT* in this branch. 

Hence the first commit in this branch, namely https://github.com/OPM/opm-simulators/pull/5620/commits/b40f5ec9a2eedbe96876097fd10e36681c455e28 cannot apply because the file it is trying to patch does not exist.

hakonhagland commented:
jenkins build this please

hakonhagland commented:
jenkins build this please

hakonhagland commented:
> Rebasing spawn_slaves on master, then this PR on spawn slaves, all seems fine to me

@akva2 Yes that worked! Thank you.

> I would start by extracting some refactorizations from this branch. Splitting AdaptiveTimeStepping.hpp into an impl file is bound to create conflicts if such a change lives long. This can be done completely separately from the other work in this PR, and will make rebasing this so much easier on you.

I see that you already submitted PR #5834. Thank you :) this will help yes. I'll wait until that is merged and then do a rebase.

hakonhagland commented:
jenkins build this please

hakonhagland commented:
> Which one should I review then?

@blattms Thanks for looking at this. The plan was to try to get this one merged and then delete #5453 due to the rebase problems, but now as I was able to rebase it we might start with #5453 ? Then, I'll remove the draft mode for that one

blattms commented:
jenkins build this serial please

blattms commented:
benchmark please

blattms reviewed (commented):
I needed to continue the review here, because this is where the fixes (also to the other PR) seem to be.

A few more rather minor comments. Overall this looks good but it also increases the complexity even more. Is the Substepper new?

Is there some documentation about what we are achieving in the time stepper? Where are the synchronization points exactly and how so we accomlish it? Subset of all the report steps ordered by time, maybe?

This should be documented in the PR and code. I could have missed it. In that case I apologize,

I need to think a bit about it and test it a bit and probably might come back with some design comments tomorrow (either for this PR or for later).

blattms commented:
A few additional questions:
- What happens if one of slaves would crash or abort?
- What happens of someone starts a simulation using the --slave parameter

hakonhagland commented:
jenkins build this please

hakonhagland commented:
jenkins build this please

hakonhagland commented:
jenkins build this please

hakonhagland commented on opm/simulators/flow/SimulatorFullyImplicitBlackoil.hpp:
```diff
@@ -208,8 +232,69 @@ class SimulatorFullyImplicitBlackoil : private SerializableSim
+        }
+        return false;
+    }
+#endif
+
+#ifdef RESERVOIR_COUPLING_ENABLED
```
Yes right! I suggest that we keep it like this to make it easier to insert other functions not related to reservoir coupling between `#endif` and the following `#if`

hakonhagland commented on opm/simulators/timestepping/AdaptiveTimeStepping.hpp:
```diff
@@ -20,12 +20,23 @@
 #include <opm/simulators/timestepping/TimeStepControl.hpp>
 #include <opm/simulators/timestepping/TimeStepControlInterface.hpp>
 
-#include <cmath>
```
I don't remember. Since the jenkins build succeeds `cmath` is indeed not needed here.

hakonhagland commented:
> What happens if one of slaves would crash or abort?

@blattms I think that would be detected by the master. Tested it now by purposely creating a segmentation fault in a slave. It was immediately noticed by the master, which called MPI_Abort() with: _mpirun noticed that process rank 1 with PID 0 on node hakon-Precision-3570 exited on signal 11 (Segmentation fault)._ Similarly, I think the master will abort the slaves and itself if any slave exits without calling `MPI_Finalize()` first (for example, due to an uncatched exception..). However, if a slave exits normally using e.g. `MPI_Finalize(); exit(0);` the master will not detect it unless it later tries to communicate with the terminated slave, in which case the error handler in the master should be called. I guess it can try to check the error code there to determine if the slave has terminated normally?

hakonhagland commented:
> What happens of someone starts a simulation using the --slave parameter

@blattms First thing to notice is that the parameter is hidden so it will not be known to the user that it exists, second if the user by accident gives the `--slave` parameter it will be detected here that the simulation is not part of a reservoir coupling run:

https://github.com/OPM/opm-simulators/blob/157cf974fa824ed14aeb63fd75880ca282b6bc72/opm/simulators/flow/ReservoirCouplingSlave.cpp#L50

and the simulation will be aborted.

blattms commented:
jenkins build this serial please

blattms commented:
Sorry, but building without MPI does not work. This prevents merging.

hakonhagland commented:
jenkins build this serial please

hakonhagland commented:
> Is there some documentation about what we are achieving in the time stepper? Where are the synchronization points exactly and how so we accomlish it? 

@blattms Good idea. I have added a description in the latest commit.

hakonhagland commented:
jenkins build this please

hakonhagland commented:
> Is there some documentation about what we are achieving in the time stepper?

@blattms See also #5893.

blattms commented:
jenkins build this serial please
